### PR TITLE
feat: remove upgrade series/base

### PR DESCRIPTION
**Pull request by SimonRichardson**
This changeset removes all of the upgrade series/base work. We're not going to transition this work over to 4.0, so we can just remove it. This may be put back in the future, but it won't be done for the initial release of 4.0.

The additional benefit to the removal of the work is that it does simpify the machine and charm epics.

Work to ensure that 3.6 agent doesn't crash when faced with missing charms will also need to be done. That work will be done against the 3.6 branch.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Migration should succeed:

```sh
$ juju bootstrap lxd src
$ juju add-model moveme
$ juju deploy ubuntu
$ juju bootstrap lxd dst
$ juju migrate src:moveme dst
```

Upgrade base should no longer be present:

```sh
$ juju upgrade-base
ERROR juju: "upgrade-base" is not a juju command. See "juju --help".

Did you mean:
        upgrade-model
```

## Documentation changes

@tmihoc Removal of update-base/series along with their hooks.

## Links

**Jira card:** [JUJU-6330](https://warthogs.atlassian.net/browse/JUJU-6330)

[JUJU-6330]: https://warthogs.atlassian.net/browse/JUJU-6330?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by hmlanigan** [comment]
@SimonRichardson Let's remove `set-application-base` as well. It was implemented for upgrade-machine and won't work with single base charms.

**Comment by SimonRichardson** [comment]
@hmlanigan amazing, was about to ask. Thanks.

File: internal/worker/uniter/remotestate/watcher_test.go [hunk 1]
```diff
@@ -246,7 +246,6 @@ func (s *WatcherSuiteIAAS) TestInitialSnapshot(c *gc.C) {
 		Relations:               map[int]remotestate.RelationSnapshot{},
 		Storage:                 map[names.StorageTag]remotestate.StorageSnapshot{},
 		ActionChanged:           map[string]int{},
-		UpgradeMachineStatus:    model.UpgradeSeriesNotStarted,
```
Comment (hmlanigan) [inline internal/worker/uniter/remotestate/watcher_test.go:249]: More to remove in this test file:

* ./internal/worker/uniter/remotestate/watcher_test.go:285:	if s.uniterClient.unit.upgradeSeriesWatcher != nil {
* ./internal/worker/uniter/remotestate/watcher_test.go:286:		s.uniterClient.unit.upgradeSeriesWatcher.changes <- struct{}{}
* ./internal/worker/uniter/remotestate/watcher_test.go:319:		s.uniterClient.unit.upgradeSeriesWatcher.changes <- struct{}{}
* ./internal/worker/uniter/remotestate/watcher_test.go:546:		s.uniterClient.unit.upgradeSeriesWatcher.changes <- struct{}{}
* ./internal/worker/uniter/remotestate/watcher_test.go:1129:		s.uniterClient.unit.upgradeSeriesWatcher.changes <- struct{}{}

File: internal/worker/uniter/remotestate/mock_test.go [hunk 2]
```diff
@@ -290,22 +290,10 @@ func (u *mockUnit) WatchRelations() (watcher.StringsWatcher, error) {
 	return u.relationsWatcher, nil
 }
 
-func (u *mockUnit) WatchUpgradeSeriesNotifications(context.Context) (watcher.NotifyWatcher, error) {
```
Comment (hmlanigan) [inline internal/worker/uniter/remotestate/mock_test.go:293]: More to remove in this test file:

* ./internal/worker/uniter/remotestate/mock_test.go:229:	upgradeSeriesWatcher             *mockNotifyWatcher

File: apiserver/facades/client/application/backend.go [hunk 3]
```diff
@@ -150,8 +150,6 @@ type Machine interface {
 	HardwareCharacteristics() (*instance.HardwareCharacteristics, error)
 	Id() string
 	PublicAddress() (network.SpaceAddress, error)
-	IsLockedForSeriesUpgrade() (bool, error)
```
Comment (hmlanigan) [inline apiserver/facades/client/application/backend.go:153]: Also remove UpdateApplicationBase from the Application interface.

File: internal/upgrades/upgradevalidation/validation_test.go [hunk 4]
```diff
@@ -129,30 +129,6 @@ func (s *upgradeValidationSuite) TestCheckForDeprecatedUbuntuSeriesForModel(c *g
 	c.Assert(blocker.Error(), gc.Equals, `the model hosts 1 ubuntu machine(s) with an unsupported base. The supported bases are: ubuntu@24.04, ubuntu@22.04, ubuntu@20.04`)
 }
 
-func (s *upgradeValidationSuite) TestGetCheckUpgradeSeriesLockForModel(c *gc.C) {
```
Comment (hmlanigan) [inline internal/upgrades/upgradevalidation/validation_test.go:132]: More references in this file to remove:

- 38:		*upgradevalidation.NewBlocker("unexpected upgrade series lock found"),
- 43:			*upgradevalidation.NewBlocker("unexpected upgrade series lock found"),
- 51:- unexpected upgrade series lock found
- 53:- unexpected upgrade series lock found
- 56:- unexpected upgrade series lock found
- 59:- unexpected upgrade series lock found
- 62:- unexpected upgrade series lock found
- 103:			return upgradevalidation.NewBlocker("unexpected upgrade series lock found"), nil
- 112:- unexpected upgrade series lock found`[1:])

**Review by hmlanigan — COMMENTED** [review]
Found more references:

-  ./cmd/containeragent/unit/package_test.go:208:		"worker/uniter/upgradeseries",
- state/application.go func (a *Application) UpdateApplicationBase(newBase Base, force bool) (err error) {
- ./apiserver/facades/client/application/backend.go:107:	UpdateApplicationBase(state.Base, bool) error
- ./internal/worker/instancemutater/mutater.go:215:			m.logger.Errorf("cannot upgrade machine-%s lxd profiles: %s", m.id, retErr.Error())
- ./internal/worker/instancemutater/mutater.go:216:			if err := m.machineApi.SetModificationStatus(ctx, status.Error, fmt.Sprintf("cannot upgrade machine's lxd profile: %s", retErr.Error()), nil); err != nil {

**Review by manadart — APPROVED** [review]
I think you've caught it all here.

**Comment by SimonRichardson** [comment]
> The others should probably be adjusted with a note about legacy behaviour - we could remove the pinning stuff altogether, but let's kick that decision down the line a bit.

I think that is something that we should tackle as an independent PR, one that can be thoroughly thought through.

**Pull request merged**